feature(pkg): dune pkg outdated

Alizter opened the pull request:
We add a `dune pkg outdated` package that will display outdated package dependencies with respect to the configured opam repositories.

It comes with a --transitive flag that shows all outdated transitive deps.

- fix #8031 

<details>
<summary>Screenshots</summary>


![image](https://github.com/ocaml/dune/assets/8614547/cf4bd490-4c5d-4152-8f22-c4fb92e3079f)
![image](https://github.com/ocaml/dune/assets/8614547/90a6df94-5e7d-4d83-8be4-7cf9d88a759e)
![image](https://github.com/ocaml/dune/assets/8614547/cabb2829-60d6-43e9-bafb-f4933cbe537b)


</details>

Alizter committed:
feature(pkg): dune pkg outdated

We add a `dune pkg outdated` package that will display outdated package
dependencies with respect to the condigured opam repositories.

It comes with a --transitive flag that shows all outdated transitive
deps.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter commented on test/blackbox-tests/test-cases/pkg/outdated.t:
```diff
@@ -0,0 +1,53 @@
+
+When printing both successes and failures, the errors should appear afterwards.
+  $ outdated --transitive
+  Error: bar is not available in any of the configured repositories
```
I changed some things around and I missed that this has now swallowed its output. Will fix.

rgrinberg commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -101,10 +101,10 @@ module Context_for_dune = struct
   ;;
 
-  type rejection = Unavailable
+  type rejection = [ `Unavailable ]
```
Why is this changed?

Alizter replied:
I had tried something but reverted it later, looks like I missed this.

rgrinberg commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -101,10 +101,10 @@ module Context_for_dune = struct
 
   let pp_rejection f = function
-    | Unavailable -> Fmt.string f "Availability condition not satisfied"
+    | `Unavailable -> Fmt.string f "Availability condition not satisfied"
```
Note for later: stop using the `Fmt` library in our code.

rgrinberg commented:
I like how the output can separate transitive vs. immediate dependencies, but I don't think this is something users will find obvious. So how about making the normal mode explain this. E.g.
```
12/15 immediate and 140/200 transitive dependencies are out of date. Displaying immediate dependencies only, add `--transitive` to display everything.
```

rgrinberg commented on bin/pkg.ml:
```diff
@@ -505,6 +518,129 @@ module Lock = struct
+                    (Dune_lang.Package_name.to_string pkg.info.name)
+                ])
+        in
+        match errors with
```
`Fiber.reraise_all`

Alizter commented on bin/pkg.ml:
```diff
@@ -505,6 +518,129 @@ module Lock = struct
+                ]
+            (* If the package is missing from the repositories, we print an error. *)
+            | Error `Package_not_found ->
+              User_error.raise
```
Rather than catching exceptions I've opted to just use `Result` on which I can manually match and filter and then raise errors later when I need to.

Alizter commented:
@rgrinberg I've refactored the error handling to use result instead rather than messing around with fibers. At the end I use `Fiber.parallel_iter` to raise all of them as user errors. I could combine this into a single raise if that's better.

I've also added support for the `--all-contexts` arg and updated the test so that it works for lockfiles in different contexts also. We also now print which lock file we are talking about to be clearer to the user.

I'll see what I can do for adding more information about immediate vs transitive deps in the message and also some counting.

Alizter commented:
I've merged the errors into a single error message, however they are still separate errors per context. See the test for how this looks in practice.

Alizter commented:
I've pushed a change that gives informative statuses based on the information present. Notably there is now a module `Dune_pkg.Better_candidate` in charge of dealing with all this status printing business. It was getting too long and complex to include in pkg.ml so I opted to move it to its separate module where its internal workings are hidden.

This also lets us test the messages extensively, so I have gone ahead and done that and enumerated all the combinations of 4 packages both transitive and immediate to test how the message ought to change.

I haven't yet settles on the forms of the status, notably I didn't incorporate @rgrinberg's suggestion of counts in the statuses yet.

I also have some cleanup to do in the cram test since there is some redundant information with the context.

I will iterate on this tomorrow.

Alizter commented:
I've pushed some more changes.

The number of packages that need updating is now reported. Have a look at the screenshots spoiler in the original post.

rgrinberg commented on src/dune_pkg/better_candidate.mli:
```diff
@@ -0,0 +1,21 @@
+  ; pkg : Lock_dir.Pkg.t
+  }
+
+val outdated_packages : t list -> (bool * Import.Package_name.t * string * string) list
```
It would be easier to read if the tuple was replaced with a record.

rgrinberg commented on src/dune_pkg/better_candidate.mli:
```diff
@@ -0,0 +1,21 @@
+open Import
```
A comment about the purpose of this module would be useful

rgrinberg commented on src/dune_pkg/better_candidate.mli:
```diff
@@ -0,0 +1,21 @@
+  }
+
+val outdated_packages : t list -> (bool * Import.Package_name.t * string * string) list
+val packages_that_were_not_found : t list -> Import.Package_name.t list
```
`Import` should be opened

rgrinberg commented on src/dune_pkg/better_candidate.mli:
```diff
@@ -0,0 +1,21 @@
+
+val outdated_status
+  :  transitive:bool
+  -> Path.Source.t
```
It's not clear what the `Path.Source.t` argument is

Alizter replied:
It's the lock directory. I've cleared this up now.

rgrinberg commented on src/dune_pkg/opam_repo.mli:
```diff
@@ -22,6 +22,7 @@ val of_opam_repo_dir_path
   -> t
 
 val repo_id : t -> Repository_id.t option
+val source : t -> string option
```
We don't have a type for sources?

Alizter replied:
We do not, this is probably why this was hidden to begin with.

rgrinberg commented on src/dune_pkg/opam_solver.mli:
```diff
@@ -22,3 +22,27 @@ val solve_lock_dir
+val better_candidate
+  :  solver_env:Solver_env.t
+  -> repos:Opam_repo.t list
+  -> version_preference:Version_preference.t
```
Why is version preference being passed? The `Better_candidate.t` generated should be independent from this option used to direct the solver. In fact, I'm not sure why the solver environment is needed either.

Alizter replied:
This is because we are using the `candidates` function from the `CONTEXT` we pass to 0install. Unfortunately this requires us to initialize some dummy parameters.

Version preference is still something that is relevant for us however as it suggests what a better candidate will be. I suppose this will always be newer packages.

 Solver env could probably be factored out but I didn't want to touch the 0install context code too much. Happy to do this if you prefer.

Alizter replied:
I've removed all the usage of 0install now.

rgrinberg commented on test/expect-tests/dune_pkg/dune_pkg_better_candidate_test.ml:
```diff
@@ -0,0 +1,360 @@
+    })
+;;
+
+let find_color_of_line line =
```
How about you just strip the colors from the output in these tests? It doesn't seem like it's worth testing.

Alizter replied:
I included them so that we can see exactly what colours the messages are, as there are a lot of them.

Alizter replied:
Its also the kind of thing that we will never see during development but users will see all the time. I therefore think its important to have some idea of what users will end up seeing.

rgrinberg commented on bin/pkg.ml:
```diff
@@ -491,6 +505,148 @@ module Lock = struct
+    let common = Common.forbid_builds common in
+    let config = Common.init common in
+    Scheduler.go ~common ~config
+    @@ fun () ->
```
Could you move the argument to `Scheduler.go` to its own function?

Alizter commented:
@rgrinberg I've moved some stuff around and removed the usage of 0install. The old `Better_candidate` module was renamed to `Outdated_packages` and the new API allows you to `find`, `pp` and `raise_errors` that occurred when searching. Some backdoors for testing were left in a `For_tests` module.

gridbugs commented on src/dune_pkg/outdated_packages.mli:
Thank you for documenting this file so thoroughly.

rgrinberg commented on src/dune_pkg/outdated_packages.ml:
```diff
@@ -0,0 +1,226 @@
+      OpamFile.OPAM.depends opam_file
+      |> OpamFilter.filter_deps
+           ~build:true
+           ~post:true
```
Should not include post dependencies

rgrinberg commented on src/dune_pkg/outdated_packages.mli:
```diff
@@ -0,0 +1,50 @@
+val raise_errors
+  :  default:User_message.Style.t Pp.t list
+  -> t
+  -> User_message.Style.t Pp.t list
```
Why does have a function that raises an error have a non polymorphic return type?

Alizter replied:
I've mixed up some concerns here. It is printing spaces when there are no errors.

rgrinberg commented on src/dune_pkg/outdated_packages.ml:
```diff
@@ -0,0 +1,226 @@
+;;
+
+(* Helpers for writing messages irregardless of the number of packages. *)
+let the_amount_out_of total amount =
```
You can just write `[x/y]` in every single one of these cases.

rgrinberg commented on src/dune_pkg/outdated_packages.ml:
```diff
@@ -0,0 +1,226 @@
+  (* If there are no outdated transitive deps then everything is up to date. *)
+  | `No_transitive_deps_outdated ->
+    [ Pp.tag User_message.Style.Success
+      @@ Pp.textf "%s is up to date!" (Path.Source.to_string lock_dir_path)
```
can you remove the exclamation mark?

rgrinberg commented on src/dune_pkg/outdated_packages.ml:
```diff
@@ -0,0 +1,226 @@
+    ]
+  | `All_transitive_deps_outdated -> packages_in_lockdir_are true number_of_outdated_deps
+  | `Some_transitive_deps_outdated ->
+    packages_in_lockdir_are false number_of_outdated_deps
```
without a label, this `false` argument is hard to read.

rgrinberg commented on src/dune_pkg/outdated_packages.ml:
```diff
@@ -0,0 +1,226 @@
+       @ default)
+;;
+
+let find ~repos ~local_packages { Lock_dir.packages; _ } =
```
If this function only needs the packages, it would be better if it just took the packages as an argument.

rgrinberg commented:
All the pretty printing code should be moved from `dune_pkg` and into the binary. All of that stuff are implementation details of the command itself.

Alizter commented:
@rgrinberg Can I propose splitting up `pkg.ml` then? Otherwise that command is getting too long.

rgrinberg commented:
How about you just add your own command as `pkg_outdated` and split the existing stuff in a subsequent PR?

Alizter commented:
@rgrinberg Sorry, I didn't mean splitting all the other commands, I just meant outdated.

Alizter commented on bin/pkg.ml:
```diff
@@ -504,6 +510,104 @@ module Lock = struct
+        let default_spacing =
+          if per_context_index < List.length per_context - 1 then [ Pp.nop ] else []
+        in
+        Console.print (Outdated_packages.raise_errors results ~default:default_spacing))
```
Sure. I'll do all the printing at once. The only issue is how to have the errors show for each context separately.

Alizter replied:
I've accumulated both the messages and the errors and now print and raise at the end.

Alizter commented on bin/pkg.ml:
```diff
@@ -529,6 +534,134 @@ module Lock = struct
+
+  let info =
+    let doc = "Check for outdated packages" in
+    let man =
+      [ `S "DESCRIPTION"
+      ; `P
+          "List packages in from lock directory that have newer versions available. By \
+           default, only direct dependencies are checked. The $(b,--transitive) flag can \
+           be used to check transitive dependencies as well."
+      ; `P "For example:"
+      ; `Pre "    \\$ dune pkg outdated"
+      ; `Noblank
+      ; `Pre "    Outdated packages in dune.lock:"
+      ; `Noblank
+      ; `Pre "    ocaml 5.1.1 < 5.2.0"
+      ; `Noblank
+      ; `Pre "    dune 3.7.1 < 3.11.0"
+      ]
```
This description needs updating.

rgrinberg commented on test/expect-tests/dune_pkg/dune:
```diff
@@ -1,5 +1,6 @@
 (library
  (name dune_pkg_unit_tests)
+ (modules dune_pkg_unit_tests)
```
don't think you need this

rgrinberg commented on test/expect-tests/dune_pkg_outdated/dune:
```diff
@@ -0,0 +1,19 @@
+(library
+ (name dune_pkg_outdated_test)
+ (modules dune_pkg_outdated_test)
```
not necessary

rgrinberg commented on test/expect-tests/dune_pkg_outdated/dune:
```diff
@@ -0,0 +1,19 @@
+  stdune
+  dune_pkg_outdated
+  dune_console
+  threads.posix
```
where do you use threads?

Alizter replied:
I don't, I got confused it confused with the other transitive deps.

Alizter commented:
@rgrinberg I've removed the
